rpc: work-around an upstream libevent bug

theuni opened the pull request:
A rare race condition may trigger while awaiting the body of a message.

This may fix some reported rpc hangs/crashes.

This work-around mimics what libevent does internally once a write has started, which is what usually happens, but not always due to the processing happening on a different thread: https://github.com/libevent/libevent/blob/e7ff4ef2b4fc950a765008c18e74281cdb5e7668/http.c#L373

Fixed upstream at: https://github.com/libevent/libevent/commit/5ff8eb26371c4dc56f384b2de35bea2d87814779

theuni committed:
rpc: work-around an upstream libevent bug

A rare race condition may trigger while awaiting the body of a message, see
upsteam commit 5ff8eb26371c4dc56f384b2de35bea2d87814779 for details.

This may fix some reported rpc hangs/crashes.

gmaxwell commented:
utACK

gmaxwell commented:
utACK.

achow101 commented:
utACK

dcousens reviewed (approved):
LGTM

promag commented:
utACK 6b58360.

For `http_set_gencb` there is no need enable/disable since the reply is sent right away in the callback:
https://github.com/bitcoin/bitcoin/blob/1b8c88451b0554502435d3883c528ad0aad1b09b/src/httpserver.cpp#L291-L296

ryanofsky commented:
I'm trying to reproduce the bug, but not having luck so far. I'm building 6b58360f9b64eb0b680a662fdfd590e47f115f44 with the fix commit reverted. My `event_get_version_number()` is 0x2001500 and I'm running

```sh
src/bitcoind -regtest
```
with

```sh
while true; do curl -s -XGET --data-binary . 127.0.0.1:18443; done
```
I also tried the python script at https://0bin.net/paste/XYtwt16siRfv1d2k#h-uIU5nDfZM34g6WvosPRu4HvNbIPzskw03AWL5i7F6

TheBlueMatt commented on src/httpserver.cpp:
```diff
@@ -601,8 +612,21 @@ void HTTPRequest::WriteReply(int nStatus, const std::string& strReply)
     evbuffer_add(evb, strReply.data(), strReply.size());
-    HTTPEvent* ev = new HTTPEvent(eventBase, true,
-        std::bind(evhttp_send_reply, req, nStatus, (const char*)nullptr, (struct evbuffer *)nullptr));
+    auto req_copy = req;
```
Why do we now need to copy req where we weren't previously?

theuni replied:
std::bind copied in req by value, but passing "this" into the lambda and using req would use this->req in the callback, which would blow up because req is set to nullptr a few lines below.

This is clunky, but I'm not sure how else to pass a class member into a lambda by value in c++11.

TheBlueMatt replied:
If the lambda were changed to simply say "req" (not "req&") then it is copied by value (in this case the pointer is copied).

theuni replied:
Class members need to be captured with 'this', so unfortunately that doesn't work.

TheBlueMatt reviewed (commented):
Looks to at least mimic the fix at https://github.com/libevent/libevent/pull/567

theuni commented on src/httpserver.cpp:
```diff
@@ -239,6 +240,16 @@ static std::string RequestMethodString(HTTPRequest::RequestMethod m)
 static void http_request_cb(struct evhttp_request* req, void* arg)
 {
+    // Disable reading to work around a libevent bug, fixed in 2.2.0.
+    if (event_get_version_number() < 0x02020001) {
```
Unsure if they'll backport, but it shouldn't hurt anything if they do. The reason for the version check is in case the behavior changes drastically in the future and enabling the read/write may be harmful. 2.2.0 isn't released, but the fix is in master.

TheBlueMatt commented:
utACK 6b58360f9b64eb0b680a662fdfd590e47f115f44, assuming https://github.com/libevent/libevent/pull/567 gets merged upstream.

theuni committed:
rpc: further constrain the libevent workaround

The bug was introduced in 2.1.6-beta, versions before that don't need the
workaround.

theuni commented:
Thanks to @ryanofsky for the version info. I bisected the upstream bug and found that it was introduced in 2.1.6-beta, so the workaround has been constrained to the appropriate range.

jonasschnelli commented:
utACK 97932cd2689659addfbb58dc6148928b73af3bd0
A squash of the two commits would be preferable IMO.

laanwj commented:
utACK 97932cd
